ECR manifest examples

rothgar opened the pull request:
Issue #171

Description of changes: Add examples for ECR CRD

Open to discussion about where the examples live and what should be in the README. I can also link to the docs about getting started so users can setup IAM and deploy the controller with CRD spec first.

rothgar commented:
Sounds good. I can move the files and update the docs tomorrow. Can you double check that the examples are correct? I'm not sure I got the tags structure right.

jaypipes commented on examples/ecr/ecr-full-repo.yaml:
```diff
@@ -0,0 +1,21 @@
+        value: "true"
+    - tag2:
+        key: "is-scanned"
+        value: "true"
```
Ah, no, this is slightly off. For ECR, the Tags field is `[]*Tag`, where `Tag` is a struct with two string fields, Key and Value:
```json
    "Tag":{
      "type":"structure",
      "members":{
        "Key":{"shape":"TagKey"},
        "Value":{"shape":"TagValue"}
      }
    },
    "TagKey":{"type":"string"},
    "TagKeyList":{
      "type":"list",
      "member":{"shape":"TagKey"}
    },
    "TagList":{
      "type":"list",
      "member":{"shape":"Tag"}
    },
```

https://github.com/aws/aws-sdk-go/blob/20cd465df9cd1c9f5c941d7843d29704b9510751/models/apis/ecr/2015-09-21/api-2.json#L1570-L1585

So, the above should be this instead:

```yaml
    tags:
    - key: "is-encrypted"
      value: "true"
    - key: "is-scanned"
      value: "true"

jaypipes commented on examples/ecr/ecr-repo.yaml:
```diff
@@ -0,0 +1,11 @@
+  tags:
+    - tag1:
+        key: "is-encrypted"
+        value: "false"
```
Should be:

```yaml
  tags:
    - key: "is-encrypted"
      value: "false"

mhausenblas commented:
Excellent work, thanks @rothgar. Since these are end-user related docs they should be located in [docs/contents/user-docs](https://aws.github.io/aws-controllers-k8s/user-docs/usage/). Can you please move it there?

rothgar commented:
@mhausenblas Do you think the full samples directory should move into docs/contenst/user-docs or just the README explaining how to deploy them?

rothgar commented:
I also see the [current install doc](https://aws.github.io/aws-controllers-k8s/user-docs/install/#static-kubernetes-manifests) suggest looking at the releases for manifests. I'm not sure that makes sense if we want to provide multiple examples for each controller so maybe examples + readme in the services/*/config/samples directory makes more sense to allow people to deploy one controller at a time.

We could mention there are examples in the services directory in the user docs but I would guess only more advanced users or developers would want to see/modify the samples per service.

rothgar commented:
I could update the usage docs with the basic examples on how to modify the resources with this PR

https://github.com/aws/aws-controllers-k8s/blob/main/docs/contents/user-docs/usage.md

mhausenblas commented:
> Do you think the full samples directory should move into docs/contenst/user-docs or just the README explaining how to deploy them?

@rothgar as per our tenets, we want to give users the best experience and that includes a consistent docs space. Based on this, *all* documentation should be under ` docs/contents/`, at the very least the entry point, for discovery purposes (mkdocs indexes everything an makes it searchable). So, one way to go about it is (shown in [mkdocs.yaml](https://github.com/aws/aws-controllers-k8s/blob/main/docs/mkdocs.yml), for the [MVP services](https://github.com/aws/aws-controllers-k8s/issues/22#issuecomment-644231634)):

```
...
nav:
    - Home: 'index.md'
    - User docs:
      - Install: 'user-docs/install.md'
      - Permissions: 'user-docs/permissions.md'
      - Usage:
        - Overview:  'user-docs/usage.md'
        - Services:
          - Amazon S3: 'user-docs/services/s3/index.md'
          - Amazon SNS: 'user-docs/services/sns/index.md'
          - Amazon SQS: 'user-docs/services/sqs/index.md'
          - Amazon ECR: 'user-docs/services/ecr/index.md'
          - Amazon DynamoDB: 'user-docs/services/dynamodb/index.md'
          - Amazon API Gateway V2: 'user-docs/services/agwy2/index.md'
... 
```

Please make sure to preview locally with `mkdocs serve` to verify rendering and links.

jaypipes commented on services/ecr/config/samples/encrypted-examlpe-repo.yaml:
```diff
@@ -0,0 +1,12 @@
+---
```
The name of this file is mispelled "examlpe" :)

jaypipes commented on services/ecr/config/samples/full-example-repo.yaml:
```diff
@@ -0,0 +1,19 @@
+        kmsKey: "9d319c68-60ea-4ed6-8a33-2e09086fad2d"
+    imageScanningConfiguration:
+        scanOnPush: true
+    imageTagMutability: true
```
When I apply this manifest, I'm getting the following failure:

```
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ kubectl apply -f services/ecr/config/samples/full-example-repo.yaml
The Repository "test-full-repository-from-ack" is invalid: spec.imageTagMutability: Invalid value: "boolean": spec.imageTagMutability in body must be of type string: "boolean"
```

That's because this field isn't a boolean :) It's actually a string enum, with valid values of "IMMUTABLE" and "MUTABLE".

Replacing `true` with `"true"` results in the CR being created successfully (since we don't have validating webhook yet...) but the controller logs errors like this:

```
2020-08-14T12:27:55.404Z	ERROR	controller-runtime.controller	Reconciler error	{"controller": "repository", "request": "default/test-full-repository-from-ack", "error": "InvalidParameterException: Invalid parameter at 'imageTagMutability' failed to satisfy constraint: 'Member must satisfy enum value set: [IMMUTABLE, MUTABLE]'"}
```

Changing the value to "IMMUTABLE" results in successful reconcilation.

rothgar replied:
`scanOnPush` is a boolean (unquoted) but `imageTagMutability` needs quotes? That seems like it would be a bad experience for users who don't know the code.

Should I change the examples and add a comment saying it should be quoted or do we want to change the type?

jaypipes replied:
imageTagMutability isn't a boolean, and "true" isn't a valid value. Only "IMMUTABLE" and "MUTABLE" are valid values.

jaypipes commented on services/ecr/config/samples/full-example-repo.yaml:
```diff
@@ -0,0 +1,19 @@
+    encryptionConfiguration:
+        encryptionType: KMS
+        # This key needs to exist already
+        kmsKey: "9d319c68-60ea-4ed6-8a33-2e09086fad2d"
```
Yes. If it doesn't, you'll get errors like this in the controller :)

```
020-08-14T12:37:04.090Z	ERROR	controller-runtime.controller	Reconciler error	{"controller": "repository", "request": "default/test-full-repository-from-ack", "error": "KmsException: The Amazon ECR action failed due to a KMS exception: Key 'arn:aws:kms:us-west-2:750630568209:key/9d319c68-60ea-4ed6-8a33-2e09086fad2d' does not exist (Service: AWSKMS; Status Code: 400; Error Code: NotFoundException; Request ID: 77029341-c5a0-4ed5-9a01-7e80c7354860; Proxy: null)\n{\n  RespMetadata: {\n    StatusCode: 400,\n    RequestID: \"0b5aae11-5dd2-4e0f-ab27-9b0d3e1d2a8b\"\n  },\n  KmsError: \"NotFoundException\",\n  Message_: \"The Amazon ECR action failed due to a KMS exception: Key 'arn:aws:kms:us-west-2:750630568209:key/9d319c68-60ea-4ed6-8a33-2e09086fad2d' does not exist (Service: AWSKMS; Status Code: 400; Error Code: NotFoundException; Request ID: 77029341-c5a0-4ed5-9a01-7e80c7354860; Proxy: null)\"\n}"}
```

jaypipes commented on services/ecr/config/samples/encrypted-examlpe-repo.yaml:
```diff
@@ -0,0 +1,12 @@
+        encryptionType: AES256
+    tags:
+    - key: "is-encrypted"
+      value: "true"
```
This manifest is all good:

```
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ kubectl apply -f services/ecr/config/samples/encrypted-examlpe-repo.yaml 
repository.ecr.services.k8s.aws/test-encrypted-repository-from-ack created
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ aws ecr describe-repositories --output table
----------------------------------------------------------------------------------------------------------
|                                          DescribeRepositories                                          |
+--------------------------------------------------------------------------------------------------------+
||                                             repositories                                             ||
|+----------------+-------------------------------------------------------------------------------------+|
||  createdAt     |  1597408711.0                                                                       ||
||  registryId    |  750630568209                                                                       ||
||  repositoryArn |  arn:aws:ecr:us-west-2:750630568209:repository/test-encrypted-repository-from-ack   ||
||  repositoryName|  test-encrypted-repository-from-ack                                                 ||
||  repositoryUri |  750630568209.dkr.ecr.us-west-2.amazonaws.com/test-encrypted-repository-from-ack    ||
|+----------------+-------------------------------------------------------------------------------------+|
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ kubectl delete -f services/ecr/config/samples/encrypted-examlpe-repo.yaml 
repository.ecr.services.k8s.aws "test-encrypted-repository-from-ack" deleted
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ aws ecr describe-repositories --output table
----------------------
|DescribeRepositories|
+--------------------+
```

jaypipes commented on services/ecr/config/samples/example-repo.yaml:
```diff
@@ -0,0 +1,10 @@
+  repositoryName: "test-repository-from-ack"
+  tags:
+  - key: "is-encrypted"
+    value: "false"
```
This manifest is all good:

```
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ git fetch --all && git fetch upstream pull/172/head:test-172
Fetching origin
Fetching upstream
remote: Enumerating objects: 17, done.
remote: Counting objects: 100% (17/17), done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 17 (delta 6), reused 15 (delta 4), pack-reused 0
Unpacking objects: 100% (17/17), done.
From github.com:aws/aws-controllers-k8s
 * [new ref]         refs/pull/172/head -> test-172
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ git checkout test-172 
Switched to branch 'test-172'
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ make build-ack-generate 
go build -tags codegen -o bin/ack-generate cmd/ack-generate/main.go
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ ./scripts/build-controller.sh ecr
Building Kubernetes API objects for ecr
Generating deepcopy code for ecr
Generating custom resource definitions for ecr
Building service controller for ecr
Generating RBAC manifests for ecr
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ source ~/.aws/role 
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ ./scripts/kind-build-test.sh -p -s ecr -r "$ROLE_ARN"
Using Kubernetes kindest/node:v1.16.9@sha256:7175872357bc85847ec4b1aba46ed1d12fa054c83ac7a8a11f5c268957fd5765
Creating k8s cluster using "kind" ...
Created k8s cluster using "kind"
Building ecr docker image
Building 'ecr' controller docker image with tag: ack-ecr-controller:ab6b049
sha256:69feab81d8bac0ffc4a5fd01c56821f754fab75e237fb26e77cca46a1adc2d3c
Loading the images into the cluster
Image: "ack-ecr-controller:ab6b049" with ID "sha256:69feab81d8bac0ffc4a5fd01c56821f754fab75e237fb26e77cca46a1adc2d3c" not yet present on node "test-f0c3a5e7-worker", loading...
Image: "ack-ecr-controller:ab6b049" with ID "sha256:69feab81d8bac0ffc4a5fd01c56821f754fab75e237fb26e77cca46a1adc2d3c" not yet present on node "test-f0c3a5e7-control-plane", loading...
Loading CRD manifests for ecr into the cluster
customresourcedefinition.apiextensions.k8s.io/repositories.ecr.services.k8s.aws created
Loading RBAC manifests for ecr into the cluster
clusterrole.rbac.authorization.k8s.io/ack-controller-role created
clusterrolebinding.rbac.authorization.k8s.io/ack-controller-rolebinding created
Loading service controller Deployment for ecr into the cluster
namespace/ack-system created
deployment.apps/ack-ecr-controller created
Running aws sts assume-role --role-arn arn:aws:iam::750630568209:role/elasticache.services.k8s.aws.full, --role-session-name tmp-role-d00cc1ff  --duration-seconds 900,
Temporary credentials generated
deployment.apps/ack-ecr-controller env updated
Added AWS Credentials to env vars map
======================================================================================================
To poke around your test manually:
export KUBECONFIG=/home/jaypipes/go/src/github.com/aws/aws-controllers-k8s/scripts/../build/tmp-test-f0c3a5e7/kubeconfig
kubectl get pods -A
======================================================================================================
repository.ecr.services.k8s.aws/ack-test-smoke-ecr created
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|                                                                                     DescribeRepositories                                                                                      |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
||                                                                                        repositories                                                                                         ||
|+--------------+---------------+--------------------------------------------------------------------+---------------------+-------------------------------------------------------------------+|
||   createdAt  |  registryId   |                           repositoryArn                            |   repositoryName    |                           repositoryUri                           ||
|+--------------+---------------+--------------------------------------------------------------------+---------------------+-------------------------------------------------------------------+|
||  1597407527.0|  750630568209 |  arn:aws:ecr:us-west-2:750630568209:repository/ack-test-smoke-ecr  |  ack-test-smoke-ecr |  750630568209.dkr.ecr.us-west-2.amazonaws.com/ack-test-smoke-ecr  ||
|+--------------+---------------+--------------------------------------------------------------------+---------------------+-------------------------------------------------------------------+|
repository.ecr.services.k8s.aws "ack-test-smoke-ecr" deleted
smoke took 8 second(s)
To resume test with the same cluster use: "-c /home/jaypipes/go/src/github.com/aws/aws-controllers-k8s/scripts/../build/tmp-test-f0c3a5e7"
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ aws ecr describe-repositories
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ export KUBECONFIG=/home/jaypipes/go/src/github.com/aws/aws-controllers-k8s/scripts/../build/tmp-test-f0c3a5e7/kubeconfig
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ kubectl get repositories
No resources found in default namespace.
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ kubectl apply -f services/ecr/config/samples/example-repo.yaml
repository.ecr.services.k8s.aws/test-repository-from-ack created
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ kubectl get repositories
NAME                       AGE
test-repository-from-ack   8s
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ aws ecr describe-repositories --output table
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|                                                                                              DescribeRepositories                                                                                               |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
||                                                                                                 repositories                                                                                                  ||
|+--------------+---------------+--------------------------------------------------------------------------+---------------------------+-------------------------------------------------------------------------+|
||   createdAt  |  registryId   |                              repositoryArn                               |      repositoryName       |                              repositoryUri                              ||
|+--------------+---------------+--------------------------------------------------------------------------+---------------------------+-------------------------------------------------------------------------+|
||  1597407652.0|  750630568209 |  arn:aws:ecr:us-west-2:750630568209:repository/test-repository-from-ack  |  test-repository-from-ack |  750630568209.dkr.ecr.us-west-2.amazonaws.com/test-repository-from-ack  ||
|+--------------+---------------+--------------------------------------------------------------------------+---------------------------+-------------------------------------------------------------------------+|
```

toricls commented on services/ecr/config/samples/README.md:
```diff
@@ -0,0 +1,20 @@
+# ECR Resources
+
+The ECR controller managed EC2 Container Registry resources inside AWS.
```
nit: "Elastic Container Registry"

mhausenblas replied:
nice catch!

rothgar commented:
> > Do you think the full samples directory should move into docs/contenst/user-docs or just the README explaining how to deploy them?
> 
> ```
> ...
> nav:
>     - Home: 'index.md'
>     - User docs:
>       - Install: 'user-docs/install.md'
>       - Permissions: 'user-docs/permissions.md'
>       - Usage:
>         - Overview:  'user-docs/usage.md'
>         - Services:
>           - Amazon S3: 'user-docs/services/s3/index.md'
>           - Amazon SNS: 'user-docs/services/sns/index.md'
>           - Amazon SQS: 'user-docs/services/sqs/index.md'
>           - Amazon ECR: 'user-docs/services/ecr/index.md'
>           - Amazon DynamoDB: 'user-docs/services/dynamodb/index.md'
>           - Amazon API Gateway V2: 'user-docs/services/agwy2/index.md'
> ... 
> ```
> 
Docs changed a lot in the last week. Is this still the structure you envision we should have for where to put install instructions per service?

Currently docs are structured with
```
...
nav:
- Home:
  - Overview: 'index.md'
  - Services: 'services.md'
- User docs:
  - Install: 'user-docs/wip.md'
  - Permissions: 'user-docs/wip.md'
  - Usage: 'user-docs/wip.md'
- Contributor docs:
  - Overview: 'dev-docs/overview.md'
  - Code generation: 'dev-docs/code-generation.md'
  - Setup: 'dev-docs/setup.md'
  - Testing: 'dev-docs/testing.md'
- Community:
  - Background: 'community/background.md'
  - FAQ: 'community/faq.md'
  - Discussions: 'community/discussions.md'
  - Contributing: 'https://github.com/aws/aws-controllers-k8s/blob/master/CONTRIBUTING.md'
  - Governance: 'https://github.com/aws/aws-controllers-k8s/blob/master/GOVERNANCE.md'
...
```

I was thinking of expanding `Install` but open to other ideas

mhausenblas commented:
Yeah, the tree I have under `Services` might be better off under `Install`. Looking forward to that @rothgar (and the other comments from Jay and Tori ;)

rothgar commented:
I _think_ this update addresses the concerns. Thanks everyone for the feedback and catching all my spelling errors.

@mhausenblas LMK if you think the install docs would be good during the dev preview. We can add more examples and docs but want to see if that content works for what you had in mind.

mhausenblas commented:
Thanks a lot @rothgar! 

> LMK if you think the install docs would be good during the dev preview. We can add more examples and docs but want to see if that content works for what you had in mind.

Good point. Let's put on hold until we have something tangible here.
